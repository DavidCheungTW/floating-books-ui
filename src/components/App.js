import React, { useState, useEffect } from "react";
import { Routes, Route } from "react-router-dom";
import NavBar from "./NavBar";
import ViewBooks from "./ViewBooks";
import BookDetails from "./BookDetails";
import RegisterBook from "./RegisterBook";
import FavouriteBooks from "./FavouriteBooks";
import OrderBooks from "./OrderBooks";
import Signin from "./Signin";
import CreateAccount from "./CreateAccount";
import EmailVerify from "./EmailVerify";
import { signOut } from "firebase/auth";
import { auth } from "../config/firebase";
import Alert from "./Alert";
import getUser from "../requests/getUser";
import "../styles/app.css";

const App = () => {
  const [loginID, setLoginID] = useState();
  const [userList, setUserList] = useState([]);
  const [userId, setUserId] = useState();
  const [selectBook, setSelectBook] = useState("");
  const [alert, setAlert] = useState({ message: "", isSuccess: true });

  useEffect(() => {
    getUser(setUserList);
    if (loginID && userList.length > 0) {
      const userRows = userList.filter((a) => a.userName === loginID);
      if (userRows.length > 0) {
        setUserId(userRows[0].id);
      }
    }
  }, [loginID, userList]);

  const updateUserList = (ul) => {
    setUserList(ul);
  };

  const handleSetSelectBook = (book) => {
    setSelectBook(book);
  };

  const handleLogin = (id) => {
    setLoginID(id);
  };

  const handleLogout = () => {
    signOut(auth)
      .then(() => {
        setLoginID();
        setUserId();
      })
      .catch((error) => {
        setAlert({
          message: `Sign out fail - ${error.message}`,
          isSuccess: false,
        });
      });
  };

  return (
    <div className="App">
      <h1>Floating Books UI</h1>
      <div className="show-user-id">
        User ID: {loginID} ({userId})
      </div>
      <NavBar loginID={loginID} onLogin={handleLogin} onLogout={handleLogout} />
      {alert.message && (
        <Alert message={alert.message} isSuccess={alert.isSuccess} />
      )}
      <Routes>
        <Route
          path="/"
          element={
            <ViewBooks
              handleSetSelectBook={handleSetSelectBook}
              userId={userId}
            />
          }
        />
        <Route
          path="book-details"
          element={<BookDetails book={selectBook} />}
        />
        <Route
          path="register-book"
          element={<RegisterBook loginID={loginID} userId={userId} />}
        />
        <Route
          path="favourite-books"
          element={
            <FavouriteBooks
              handleSetSelectBook={handleSetSelectBook}
              loginID={loginID}
              userId={userId}
            />
          }
        />
        <Route
          path="order-books"
          element={
            <OrderBooks
              handleSetSelectBook={handleSetSelectBook}
              loginID={loginID}
              userId={userId}
            />
          }
        />
        <Route
          path="signin"
          element={
            <Signin
              onSetUser={handleLogin}
              userList={userList}
              updateUserList={updateUserList}
            />
          }
        />
        <Route
          path="signin/create-account"
          element={<CreateAccount onSetUser={handleLogin} />}
        />
        <Route
          path="signin/email-verify"
          element={
            <EmailVerify userList={userList} updateUserList={updateUserList} />
          }
        />
      </Routes>
    </div>
  );
};

export default App;
