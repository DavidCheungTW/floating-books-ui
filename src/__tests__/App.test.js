import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import App from "../components/App";

describe("App", () => {
  it("Renders as expected", () => {
    const rendered = renderer.create(
      <Router>
        <App />
      </Router>
    );
    expect(rendered).toMatchSnapshot();
  });

  it("Renders as expected", () => {
    render(
      <Router>
        <App />
      </Router>
    );

    const linkElement = screen.getByText(/Floating Books UI/i);
    expect(linkElement).toBeInTheDocument();
  });
});
