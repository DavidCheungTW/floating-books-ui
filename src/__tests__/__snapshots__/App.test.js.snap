// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders as expected 1`] = `
<div
  className="App"
>
  <h1>
    Floating Books UI
  </h1>
  <div
    className="show-user-id"
  >
    User ID: 
     (
    )
  </div>
  <div
    className="navbar"
  >
    <ul
      className="navbar-links"
    >
      <li
        className="navbar-links-item"
      >
        <a
          className="item"
          href="/"
          onClick={[Function]}
        >
          View Books
        </a>
      </li>
      <li
        className="navbar-links-item"
      >
        <a
          className="item"
          href="/favourite-books"
          onClick={[Function]}
        >
          Favourite Books
        </a>
      </li>
      <li
        className="navbar-links-item"
      >
        <a
          className="item"
          href="/order-books"
          onClick={[Function]}
        >
          Order Books
        </a>
      </li>
      <li
        className="navbar-links-item"
      >
        <a
          className="item"
          href="/register-book"
          onClick={[Function]}
        >
          Register a Book
        </a>
      </li>
      <li
        className="navbar-links-item signin-button"
      >
        <a
          className="item"
          href="/signin"
          onClick={[Function]}
        >
          Sign in
        </a>
      </li>
    </ul>
  </div>
  <div
    className="view-books"
  >
    <div
      className="column left"
    >
      <div
        className="side-bar"
      >
        <form
          className="search-form"
          onSubmit={[Function]}
        >
          <input
            id="searchText"
            name="searchText"
            onChange={[Function]}
            placeholder="Search by title"
            type="text"
            value=""
          />
          <button
            type="submit"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1014.64 969.04L703.71 656.207c57.952-69.408 92.88-158.704 92.88-256.208 0-220.912-179.088-400-400-400s-400 179.088-400 400 179.088 400 400 400c100.368 0 192.048-37.056 262.288-98.144l310.496 312.448c12.496 12.497 32.769 12.497 45.265 0 12.48-12.496 12.48-32.752 0-45.263zM396.59 736.527c-185.856 0-336.528-150.672-336.528-336.528S210.734 63.471 396.59 63.471c185.856 0 336.528 150.672 336.528 336.528S582.446 736.527 396.59 736.527z"
              />
            </svg>
          </button>
        </form>
      </div>
    </div>
    <div
      className="column right"
    />
  </div>
</div>
`;
